### PR TITLE
Page: Add inline rename functionality

### LINKED ISSUE #68070: Rename folder
"Inline" Folder rename from page title

Related bugs:
- [ ] When renaming the folder, the page title does not update, until I refresh
- [ ] When trying to rename the folder a second time (even after refresh, but without leaving the page), we get an error: "the folder has been changed by someone else"

### PR DESCRIPTION
**Pull request by ashharrison90**
**What is this feature?**

- creates a new `EditableTitle` component
- exposes an `onEditTitle` prop on the `Page` component
- when `onEditTitle` is present, use `EditableTitle` as the page title to provide inline rename capabilities

**Why do we need this feature?**

- better UX for individual item pages

**Who is this feature for?**

- everyone!

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/68070

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: packages/grafana-ui/src/components/AutoSaveField/AutoSaveField.tsx [hunk 1]
```diff
@@ -95,7 +96,7 @@ export function AutoSaveField<T = string>(props: Props<T>) {
         disabled={disabled}
         error={error || (fieldState.showError && saveErrorMessage)}
         ref={fieldRef}
-        className={styles.widthFitContent}
+        className={classNames(styles.widthFitContent, restProps.className)}
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply packages/grafana-ui/src/components/AutoSaveField/AutoSaveField.tsx:99]: Actually, do you even use this component here?

[the remark this answers is not in the document]
Comment (ashharrison90) [reply packages/grafana-ui/src/components/AutoSaveField/AutoSaveField.tsx:99]: oh yeah I didn't use Autosavefield in the end so can remove this change 👌

**Comment by tskarhed** [comment]
The time is time for the Input to disappear after I've blurred or pressed enter feels a little bit clunky. Is this the proper use case for AutoSaveField? I would prefer the frontend would directly flip to the edited title, and then flip back and display an error if it fails.

**Comment by ashharrison90** [comment]
@tskarhed done 👍 you're right, it feels much snappier 😍

File: public/app/core/components/Page/EditableTitle.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+  return !isEditing ? (
+    <div className={styles.textContainer}>
+      <div className={styles.textWrapper}>
+        {/*
+          use localValue if it exists
+          this is to prevent the title from flickering back to the old value after the user has edited
+          caused by the delay between the save completing and the new value being refetched
+        */}
+        <H1 truncate>{localValue ?? value}</H1>
```
Comment (joshhunt) [inline public/app/core/components/Page/EditableTitle.tsx:59]: I think there's an iffy bit here where the parent cannot change the `value` prop after its ever been edited?

I _think_ this is an instance where we want to react to changes in the `value` prop and sync it to state.

```ts
const [localValue, setLocalValue] = useState(value);
useEffect(() => {
  setLocalValue(value);
}, [value])
```

(this should not cause a re-render on first mount as the value doesn't change)

**Review by joshhunt — APPROVED** [review]
Excellent work. This is thoughly developed, tested, and thought about - well done 👏🏻 

I'm a little bit uncertain on the interaction, especially with how it's just.... saved automatically without much feedback on blur, but happy to defer to others and see what it's like in practice.

(just the two small comments)

**Comment by joshhunt** [comment]
Also, long title that wraps pushes the page actions down to a new line

image

but then when you hit edit the actions wrap back up 

image

**Pull request merged**